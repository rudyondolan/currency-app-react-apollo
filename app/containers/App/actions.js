/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import { SET_AMOUNT, SET_BASE } from './constants';

/**
 * Set amount currency action
 * @param {number} amount List of numbers that will be used for input
 */
export function setAmount(amount) {
  return {
    type: SET_AMOUNT,
    amount,
  };
}

/**
 * Set base currency action
 * @param {string} base A string currency code in 3 digits
 */
export function setBase(base) {
  return {
    type: SET_BASE,
    base,
  };
}
