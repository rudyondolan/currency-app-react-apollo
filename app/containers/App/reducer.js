/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can change our
 * application state.
 * To add a new action, add it to the switch statement in the reducer function
 *
 * Example:
 * case YOUR_ACTION_CONSTANT:
 *   return state.set('yourStateVariable', true);
 */

import { fromJS } from 'immutable';

import { SET_AMOUNT, SET_BASE } from './constants';

// The initial state of the App
const initialState = fromJS({
  base: 'USD',
  amount: 1,
});

function appReducer(state = initialState, action) {
  switch (action.type) {
    case SET_AMOUNT:
      return state.set('amount', action.amount);
    case SET_BASE:
      return state.set('base', action.base);
    default:
      return state;
  }
}

export default appReducer;
