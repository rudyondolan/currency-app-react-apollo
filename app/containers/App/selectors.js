/**
 * The global state selectors
 */

import { createSelector } from 'reselect';

const selectGlobal = state => state.get('global');

const selectRoute = state => state.get('route');

const makeSelectAmount = () =>
  createSelector(selectGlobal, globalState => globalState.get('amount'));

const makeSelectBase = () =>
  createSelector(selectGlobal, globalState => globalState.get('base'));

const makeSelectLocation = () =>
  createSelector(selectRoute, routeState => routeState.get('location').toJS());

export { selectGlobal, makeSelectBase, makeSelectAmount, makeSelectLocation };
