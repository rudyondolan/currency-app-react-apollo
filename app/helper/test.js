import React from 'react';
import { IntlProvider, intlShape } from 'react-intl';
import { mount, shallow } from 'enzyme';

const intlProvider = new IntlProvider({ locale: 'en' }, {});
const { intl } = intlProvider.getChildContext();

function nodeWithIntlProp(node) {
  return React.cloneElement(node, { intl });
}

export function shallowWithIntl(node, { context, ...additionalOptions } = {}) {
  return shallow(nodeWithIntlProp(node), {
    context: Object.assign({}, context, { intl }),
    ...additionalOptions,
  });
}

export function mountWithIntl(
  node,
  { context, childContextTypes, ...additionalOptions } = {},
) {
  return mount(nodeWithIntlProp(node), {
    context: Object.assign({}, context, { intl }),
    childContextTypes: Object.assign(
      {},
      { intl: intlShape },
      childContextTypes,
    ),
    ...additionalOptions,
  });
}
