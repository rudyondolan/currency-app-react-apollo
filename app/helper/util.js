export function toFloat(value) {
  if (typeof value === 'string') {
    const fVal = parseFloat(value);
    if (typeof fVal === 'number' && !isNaN(fVal)) {
      return fVal;
    }

    return 1;
  }

  return value;
}

