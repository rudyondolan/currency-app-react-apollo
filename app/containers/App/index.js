/**
 *
 * App
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Helmet } from 'react-helmet';
import { Switch, Route, withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';

import Header from 'components/Header';
import AboutPage from 'containers/AboutPage/Loadable';
import ExchangePage from 'containers/ExchangePage/Loadable';
import NotFoundPage from 'containers/NotFoundPage/Loadable';

import { makeSelectAmount, makeSelectBase } from './selectors';
import { setAmount, setBase } from './actions';

function App(props) {
  const { onSetAmount, onSetBase, base, amount } = props;
  return (
    <div>
      <Helmet
        titleTemplate="%s - React.js Boilerplate"
        defaultTitle="React.js Boilerplate"
      >
        <meta name="description" content="A React.js Boilerplate application" />
      </Helmet>
      <Header
        amount={amount}
        base={base}
        onSetAmount={onSetAmount}
        onSetBase={onSetBase}
      />
      <Switch>
        <Route exact path="/" component={ExchangePage} />
        <Route exact path="/about" component={AboutPage} />
        <Route path="" component={NotFoundPage} />
      </Switch>
    </div>
  );
}

App.propTypes = {
  base: PropTypes.string,
  amount: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  onSetAmount: PropTypes.func,
  onSetBase: PropTypes.func,
};

export function mapDispatchToProps(dispatch) {
  return {
    onSetAmount: amount => dispatch(setAmount(amount)),
    onSetBase: base => dispatch(setBase(base)),
  };
}

const mapStateToProps = createStructuredSelector({
  amount: makeSelectAmount(),
  base: makeSelectBase(),
});

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(
  withRouter,
  withConnect,
)(App);
