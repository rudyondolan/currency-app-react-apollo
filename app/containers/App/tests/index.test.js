import React from 'react';
import { shallowWithIntl } from 'helper/test';
import App from '../index';

describe('<App />', () => {
  test('should match with the snapshot', () => {
    const wrapper = shallowWithIntl(<App />);
    expect(wrapper).toMatchSnapshot();
  });
});
