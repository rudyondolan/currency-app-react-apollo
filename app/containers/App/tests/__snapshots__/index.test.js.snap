// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match with the snapshot 1`] = `<Route />`;
